plugins/inputs/windows_event_log/wineventlog/wineventlog_test.go

```diff
@@ -0,0 +1,163 @@
+	assert.NoError(t, elog.Close())
+	// bad name
+	elog = NewEventLog("FakeBadElogName", LEVELS, GROUP_NAME, STREAM_NAME,
+		RENDER_FMT, DEST, STATE_FILE_PATH, BATCH_SIZE, RETENTION)
+	assert.Error(t, elog.Open())
+	assert.Zero(t, elog.eventHandle)
+	// bad LEVELS does not cause Open() to fail.
+	// bad wlog.eventOffset does not cause Open() to fail.
+}
+
+// TestReadGoodSource will verify we can read events written by a registered
+// event log source.
+func TestReadGoodSource(t *testing.T) {
+	elog := NewEventLog(NAME, LEVELS, GROUP_NAME, STREAM_NAME, RENDER_FMT, DEST,
+		STATE_FILE_PATH, BATCH_SIZE, RETENTION)
+	assert.Equal(t, nil, elog.Open())
+	seekToEnd(t, elog)
+	writeEvents(t, 10, true, "CWA_UnitTest111", 777)
+	// Without a sleep we do not read any events!!!
+	time.Sleep(1 * time.Second)
+	checkEvents(t, elog, 10, "[Application] [ERROR] [777] [CWA_UnitTest111] ")
+	assert.Equal(t, nil, elog.Close())
+}
+
+// TestReadBadSource will verify that we cannot read events written by an
+// unregistered event log source.
+func TestReadBadSource(t *testing.T) {
+	elog := NewEventLog(NAME, LEVELS, GROUP_NAME, STREAM_NAME, RENDER_FMT, DEST,
+		STATE_FILE_PATH, BATCH_SIZE, RETENTION)
+	assert.Equal(t, nil, elog.Open())
+	seekToEnd(t, elog)
+	writeEvents(t, 10, false, "CWA_UnitTest222", 888)
+	time.Sleep(1 * time.Second)
+	checkEvents(t, elog, 0, "[Application] [ERROR] [888] [CWA_UnitTest222] ")
+	assert.Equal(t, nil, elog.Close())
+}
+
+// TestReadWithBothSources will verify we can read events written by a
+// registered event log source, even if the batch contains events from an
+// unregistered source too.
+func TestReadWithBothSources(t *testing.T) {
+	elog := NewEventLog(NAME, LEVELS, GROUP_NAME, STREAM_NAME, RENDER_FMT, DEST,
+		STATE_FILE_PATH, BATCH_SIZE, RETENTION)
+	assert.Equal(t, nil, elog.Open())
+	seekToEnd(t, elog)
+	writeEvents(t, 10, true, "CWA_UnitTest111", 777)
+	writeEvents(t, 10, false, "CWA_UnitTest222", 888)
+	time.Sleep(1 * time.Second)
+	checkEvents(t, elog, 10, "[Application] [ERROR] [777] [CWA_UnitTest111] ")
+	assert.Equal(t, nil, elog.Close())
+}
+
+// seekToEnd skips past all current events in the event log.
+func seekToEnd(t *testing.T, elog *windowsEventLog) {
+	// loop until we stop getting records.
+	numRetrieved := 0
+	for {
+		records := elog.read()
+		//assert.Equal(t, nil, err)
+		t.Logf("seekToEnd() current count %v", len(records))
+		numRetrieved += len(records)
+		if len(records) == 0 {
+			break
+		}
+	}
+	t.Logf("seekToEnd() total %v", numRetrieved)
+}
+
+// writeEvents writes msgCount number of events to the Application event log.
+// Optionally register the given logSrc.
+// Fail the test if an error occurs.
+func writeEvents(t *testing.T, msgCount int, doRegister bool, logSrc string, eventId uint32) {
+	if doRegister {
+		// Expected to fail if unit test previously ran and installed the event src.
+		_ = eventlog.InstallAsEventCreate(logSrc, eventlog.Info|eventlog.Warning|eventlog.Error)
```

Comment:
Do we not want to check what gets returned here?

Reply:
I don't think so.
The call should succeed the first time the test is run.
Subsequent test runs should fail and that is okay.
I don't think we need to bother with "cleaning up" the registered event source either.

For context I expect this test to be run by the GitHub actions runner which uses `windows-latest` aka Windows 2019:
https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources
And it is an actual VM, not just a docker container - so I expect this test to to work.